lcls_tools/common/data_analysis/bmad_modeling.py

```diff
@@ -0,0 +1,573 @@
+            print(f"\n{element} BMAG_X {bmag_a:3.2f}, BMAG_Y {bmag_b:3.2f}")
+        print("")
+
+    def quad_table(self, tao, pct_lim=1, show_energy=False):
+        """Display table of quad elements BDES, BMOD, Bmad model
+        BDES and  Bamd model energy.  Filter
+        by pct_lim (BMOD - Bmad model)/ BMOD"""
+        quads = tao.lat_list("quad::Q*", "ele.name", flags="-no_slaves")
+        if show_energy:
+            print("Ele.    Device            EACT    EDES    E_TOT    ")
+        else:
+            print("Ele.    Device           BDES    BMOD    Bmad     %")
+        for element in quads[1:]:
+            device = tao.ele_head(element)["alias"]
+            bmod = epics.caget(device + ":BMOD")
+            bdes = epics.caget(device + ":BDES")
+            eact = epics.caget(device + ":EACT")
+            edes = epics.caget(device + ":EDES")
+            e_tot = tao.ele_gen_attribs(element)["E_TOT"] / 1e9
+            model_bdes = self.get_bmad_bdes(tao, element)
+            if show_energy:
+                print(
+                    f"{element:7s} {device:15s} {eact:7.3f}"
+                    f"{edes:7.3f} {e_tot:7.3f}"
+                )
+            else:
+                percent = (
+                    100 * abs((bmod - model_bdes) / bmod) if bmod != 0 else 0)
+                if percent > pct_lim:
+                    print(
+                        f"{element:7s} {device:15s} {bdes:7.3f}"
+                        f"{bmod:7.3f} {model_bdes:7.3f} {percent:7.3}"
+                    )
+
+    def get_bmad_bdes(self, tao, element, b1_gradient=[]):
+        """Returns BDES from Bmad B1_GRADIENT or given gradient"""
+        ele_attr = tao.ele_gen_attribs(element)
+        if not b1_gradient:
+            b1_gradient = ele_attr["B1_GRADIENT"]
+        return -b1_gradient * ele_attr["L"] * 10
+
+    def create_emitmeas_datum(self, tao, element):
+        tao.cmd("set global lattice_calc_on = F")
+        tao.data_d2_create(f"emitmeas{element}", 1, "twiss^^1^^4")
+        tao.datum_create(
+            f"emitmeas{element}.twiss[1]",
+            "beta.a",
+            ele_name=element,
+            merit_type="target",
+            weight=10,
+        )
+        tao.datum_create(
+            f"emitmeas{element}.twiss[2]",
+            "alpha.a",
+            ele_name=element,
+            merit_type="target",
+            weight=10,
+        )
+        tao.datum_create(
+            f"emitmeas{element}.twiss[3]",
+            "beta.b",
+            ele_name=element,
+            merit_type="target",
+            weight=10,
+        )
+        tao.datum_create(
+            f"emitmeas{element}.twiss[4]",
+            "alpha.b",
+            ele_name=element,
+            merit_type="target",
+            weight=10,
+        )
+        tao.data_set_design_value()
+        tao.cmd("set global lattice_calc_on = T")
+
+    def plot_twiss(self, tao, output, info="", xoff=0):
+        fig, ax = plt.subplots(figsize=(8, 4))
```

Comment:
Same plotting comment. We can generalize and move this.

Reply:
done.